### ISSUE TITLE
*> is strict in second argument

### ISSUE THREAD

**Issue by Profpatsch** [scalaz/scalaz#1020]
```
none[Int] *> (sys.error("bang")) 
> java.lang.RuntimeException: bang
```

I’d argue it should be non-strict in the second argument, in order to make short-circuiting possible (if that is possible?)

Not sure about `<*`, whether it should be non-strict in second argument or first. Probably second.

**Comment by tonymorris** [comment]
This came up on IRC. i think it goes back to whether `Apply` should be non-strict, and then the implications for any change on that, etc.

**Comment by kevinmeredith** [comment]
Per http://stackoverflow.com/questions/36335417/make-argument-lazy-without-changing-function-signature#comment60293037_36335417, is there a sensible way to address this item?

Note - I'm **assuming** that using a `by-name` parameter is not acceptable here.

**Comment by jbgi** [comment]
By using a church encoding (so not a case class) it is possible to create both strict and lazy constructors and thus not rely on by-name parameters. I took this approach to implement algebraic data types and laziness for java: https://github.com/derive4j/derive4j#first-class-laziness (see eg. https://gist.github.com/jbgi/43c1bd0ab67e3f4b9634#file-lists-java-L99)

**Comment by tonymorris** [comment]
I have never found a good solution to this problem [in Scala], except to have every variant of strict or non-strict for every possible function argument. Of course, this is not practical, so just implement "whatever is likely to be useful for now" and if another variant comes up as required, then implement that too (etc etc).

**Comment by kevinmeredith** [comment]
For my own understanding, why can't *> simply have a by name second parameter?

Further, what harm, if any, is done by replacing strictly evaluated arguments with lazy ones?

**Comment by tonymorris** [comment]
It's a trade-off. I agree that there are benefits to have by-name
argument, but that incurs other penalties, such as not working well with
scala e.g. the first argument in object.method(args) is always strict,
and so why do we use those? Because type-inference works better, etc etc.

I have found no better solution to this problem (in scala) than to
simply have every permutation of strict/non-strict function arguments,
while omitting those that don't come up very often just due to the
workload. So a *> function that is non-strict in second argument is
definitely worthwhile, just it would have to be a different function,
and nobody has yet written it.

**Comment by S11001001** [comment]
@Profpatsch By #1318, it's likely that `*>ByName` will be introduced with the nonstrict-right behavior for 7.3.0.

**Comment by Profpatsch** [comment]
Nice. I’m just glad I don’t have to write Scala anymore.

**Comment by S11001001** [comment]
Fixed by #1318.
